<a href="https://colab.research.google.com/github/nathanianah/InformationVisualization/blob/main/lab1_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Basic data manipulation using Python `pandas`.

## Packages

- `xlrd`: A package used to load an Excel sheet to Python.  It is required by `pandas` when transforming an Excel sheet to a data frame.

- `pandas`: A library to handle table data.

In [2]:
# Install the xlrd package.  As it is not included in the Google Colab's standard set of packages.
!pip install xlrd

from IPython.core.display import display, HTML, Markdown
def hr(): display(HTML('<hr/>'))
def heading(n, title): display(HTML(f'<h{n}>{title}</h{n}>'))
def h1(title): heading(1, title)
def h2(title): hr(); heading(2, title)
def h3(title): heading(3, title)
def paragraph(content): display(HTML(f'<p>{content}</p>'))

import pandas as pd
pd.options.display.max_rows = 8

## Python list

- l = [0, 1, 2, 3]

- l = list(range(20))

- `len(l)`: list length

<hr/>

### Python list is 0-indexed: the first item in a list is referenced by `[0]`
- `l[0]`, `l[1]`, `l[9]`: first, second, and tenth items in the list named `l`

<hr/>

### Slices

- `l[0:7]`: First to the seventh items in a list

- `l[3:7]`: Forth, fifth, sixth, and seventh items in a list

- `l[7:]`: Eighth and after

- `l[1::2]`: Every one items from the second

- `l[1::3]`: Every two items from the second

- `l[3:10:2]`: Every one items from forth upto tenth item

- `l[3:10:3]`: Every two items from forth upto tenth

<hr/>

### $n$-th element to the last

- `l[-1]`: The last element

- `l[-2]`: The second to the last

- `l[-3:]`: The third to the last

- `l[-5::2]`: Every one item to the fifth the last (..., l[-9], l[-7], l[-5])

- `l[2:-2]`: Every one item from the third to the second the last

In [3]:
h2('List definition')
l = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
print(l)

l = list(range(1, 21))
print(l)

h2('List length')
print('len(l):', len(l), '\n')

h2('Referencing a list element')
print('l[0]:', l[0])
print('l[1]:', l[1])
print('l[9]:', l[9])
print('l[19]:', l[19])

h2('Referencing a subsequence of the list')
print('l[0:7]:', l[0:7])
print('l[3:7]:', l[3:7])
print('l[7:]:', l[7:], '\n')

h2('Skipping')
print('l[3:10:2]:', l[3:10:2])
print('l[3:10:3]:', l[3:10:3])
print('l[1::2]:', l[1::2])
print('l[1::3]:', l[1::3], '\n')

h2('From the end of the list')
print('l[-1]:', l[-1])
print('l[-2]:', l[-2])
print('l[-3:]:', l[-3:])
print('l[-5::2]:', l[-5::2])
print('l[2:-2]:', l[2:-2])

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]


len(l): 20 



l[0]: 1
l[1]: 2
l[9]: 10
l[19]: 20


l[0:7]: [1, 2, 3, 4, 5, 6, 7]
l[3:7]: [4, 5, 6, 7]
l[7:]: [8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20] 



l[3:10:2]: [4, 6, 8, 10]
l[3:10:3]: [4, 7, 10]
l[1::2]: [2, 4, 6, 8, 10, 12, 14, 16, 18, 20]
l[1::3]: [2, 5, 8, 11, 14, 17, 20] 



l[-1]: 20
l[-2]: 19
l[-3:]: [18, 19, 20]
l[-5::2]: [16, 18, 20]
l[2:-2]: [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]


## Loading an Excel sheet from GitHub repository

We use `pandas.read_excel` function to load a [Social Progress Index (SPI)](https://www.socialprogress.org) dataset stored in a GitHub repository.  Loading from a GitHub repository is the easiest to provide data to Google Colaboratory.  There are a few other means.  Find more [in this article](https://towardsdatascience.com/3-ways-to-load-csv-files-into-colab-7c14fcbdcb92).

This dataset contains multiple sheet.  We are interested in a sheet named `2019`.

The `pandas.read_excel` function converts the Excel sheet into a data structure called **Pandas data frame**.  A data frame offers many useful statistic analysis tools.

The `display` function from IPython package beautifully render the content of the sheet.

In [4]:
h2('Upload an Excel sheet from a GitHub repository and convert to a dataframe.')
spi2019 = pd.read_excel('https://github.com/wakita/y20-vis/blob/master/docs/data/spi2019.xlsx?raw=true',
                      sheet_name='2019')
display(spi2019)

h2('Indices')
indices = list(spi2019.keys())
print(indices)

2019 ranking  ... Percent of tertiary students enrolled in globally ranked universities
0             NaN  ...                                          13.290866                   
1           143.0  ...                                           0.000000                   
2            54.0  ...                                           0.000000                   
3            79.0  ...                                          15.000000                   
..            ...  ...                                                ...                   
235           NaN  ...                                           0.000000                   
236           NaN  ...                                           0.000000                   
237           NaN  ...                                           0.000000                   
238           NaN  ...                                           0.000000                   

[239 rows x 71 columns]

['2019 ranking', 'Country', 'Code', 'Status', 'Social Progress Index', 'Basic Human Needs', 'Foundations of Wellbeing', 'Opportunity', 'Nutrition and Basic Medical Care', 'Water and Sanitation', 'Shelter', 'Personal Safety', 'Access to Basic Knowledge', 'Access to Information and Communications', 'Health and Wellness', 'Environmental Quality', 'Personal Rights', 'Personal Freedom and Choice', 'Inclusiveness', 'Access to Advanced Education', 'Undernourishment (% of pop.)', 'Maternal mortality rate (deaths/100,000 live births)', 'Child mortality rate (deaths/1,000 live births)', 'Child stunting (% of children)', 'Deaths from infectious diseases (deaths/100,000 people)', 'Access to at least basic drinking water (% of pop.)', 'Access to piped water (% of pop.)', 'Access to at least basic sanitation facilities (% of pop.)', 'Rural open defecation (% of pop.)', 'Access to electricity (% of pop.)', 'Quality of electricity supply (1=low; 7=high)', 'Household air pollution attributable deaths (

SPI dataset compares UN member states (countries and regions) using 54 indices called SPI indices.  The data set contains data from years 2014 through 2019.

The variable `spi2019` points to a Pandas data frame object that stores the sheet named `2019` of the Excel file.  The sheet corresponds to the year 2019 record of the SPI dataset.

For the detail of the SPI indices, please [download the Excel sheet](https://github.com/wakita/y20-vis/raw/master/docs/data/spi2019.xlsx) and refer to the **Definitions** sheet.

The variable `indices` contains the list of index names.

The table contains Year 2019 SPI data for a total of 236 UN member states.

This table contains 71 columns. The first four columns give the SPI ranking, name, code, and status of the member state.  The last 51 columns (the `U` through `BS` columns in Excel sheet) are basic indicators.  The rest are abstracted indicators that are aggregated from the basic indicators. The columns are obtained by `spi2019.keys()`, which is converted to Python list and stored in `indices`.

The example below displays some of the data for the first five of these countries and regions (Afghanistan, Angola, Albania, United Arab Emirates, Argentina).

In the following examples, the first five columns of the table are shown. Note that Pandas' table corresponds to the slice notation similar to Python's list. `[:n]` means a group of $n$ elements at the top of a structure such as a list, `[n:]` means a group of elements after the nth element. Also, a negative value as the index of an element points to the nth element from the end of the list etc.

In [5]:
h2('The list of SPI indices')
print('#SPI Indices = ' + str(len(indices)) + '\n  - ' +
      '\n  - '.join(indices[:6]) +
      ',\n    ...,\n  - ' +
      '\n  - '.join(indices[-5:]))

h2('Member states')
display(spi2019['Country'])

h2('Member state code')
print(spi2019['Code'], '\n')

h2('Showing only part of the big table')
indices5 = indices[:5]  # The index names for the first four attributes
# Presenting the first five attributes of the first five member states
h3('indices5:')
print(indices5)
h3('Five indices for the five countries')
display(spi2019[indices[:5]][1:6])

h2('The number the list of member states')
print('#Member states =', len(spi2019['Country']))
display(spi2019['Country'])

#SPI Indices = 71
  - 2019 ranking
  - Country
  - Code
  - Status
  - Social Progress Index
  - Basic Human Needs,
    ...,
  - Equality of political power by social group (0=unequal power; 4=equal power)
  - Years of tertiary schooling
  - Women's average years in school
  - Globally ranked universities (points)
  - Percent of tertiary students enrolled in globally ranked universities


0               World
1         Afghanistan
2             Albania
3             Algeria
            ...      
235    Western Sahara
236             Syria
237         Venezuela
238             Yemen
Name: Country, Length: 239, dtype: object

0      WWW
1      AFG
2      ALB
3      DZA
      ... 
235    ESH
236    SYR
237    VEN
238    YEM
Name: Code, Length: 239, dtype: object 



['2019 ranking', 'Country', 'Code', 'Status', 'Social Progress Index']


2019 ranking      Country Code  Status  Social Progress Index
1         143.0  Afghanistan  AFG  Ranked                  38.60
2          54.0      Albania  ALB  Ranked                  71.57
3          79.0      Algeria  DZA  Ranked                  67.06
4         137.0       Angola  AGO  Ranked                  42.58
5          42.0    Argentina  ARG  Ranked                  76.86

#Member states = 239


0               World
1         Afghanistan
2             Albania
3             Algeria
            ...      
235    Western Sahara
236             Syria
237         Venezuela
238             Yemen
Name: Country, Length: 239, dtype: object

## Selecting a row of the data frame

You can create a mask for the rows of a table by writing a predicate (a boolean expression) over the column values. For example, the following formula would be the truth value for each member state, with Albania taking the true value and the other member states taking the false values.

> spi2019['Country'] == 'Albania'

We can apply such a truth mask to `spi2019`. That way, we can retrieve only the rows that correspond to the true values from the table.

> spi2019[spi2019['Country'] == 'Albania']

In [6]:
h2('SPI indices')
display(spi2019['Social Progress Index'])

h2('The mask that matches a member state whose name is "Albania"')
mask_Albania = spi2019['Country'] == 'Albania'
display(mask_Albania)

h2('Application of the mask to the table')
display(spi2019[mask_Albania][indices[:7]])
hr()

h2('A mask that makes indicates India, Ireland, ..., and Japan')
mask_india2japan = (spi2019['Country'] >= 'India') & (spi2019['Country'] <= 'Japan')

h2('Application of the mask to the table')
display(spi2019[mask_india2japan])
# Note that the member states are not sorted according to their names.  We find Iraq and Isle of Man after Japan

0      64.47
1      38.60
2      71.57
3      67.06
       ...  
235      NaN
236      NaN
237      NaN
238      NaN
Name: Social Progress Index, Length: 239, dtype: float64

0      False
1      False
2       True
3      False
       ...  
235    False
236    False
237    False
238    False
Name: Country, Length: 239, dtype: bool

2019 ranking  Country  ... Basic Human Needs Foundations of Wellbeing
2          54.0  Albania  ...             85.03                    79.03

[1 rows x 7 columns]

2019 ranking  ... Percent of tertiary students enrolled in globally ranked universities
63          102.0  ...                                                0.0                   
64           85.0  ...                                                5.0                   
65           86.0  ...                                               10.0                   
66           14.0  ...                                               55.0                   
..            ...  ...                                                ...                   
69           50.0  ...                                               60.0                   
70           10.0  ...                                               30.0                   
159           NaN  ...                                               40.0                   
229           NaN  ...                                                0.0                   

[10 rows x 71 columns]

## Exercise

1. Show the data for Japan and Norway

1. Show the columns **Basic Human Needs**, **Foundations of Wellbeing**, **Opportunity** for Japan and Norway

In [20]:
h2('A mask for Japan and Norway')
mask_JpnNor = (spi2019['Country'] == 'Japan') | (spi2019['Country'] == 'Norway')
display(spi2019[mask_JpnNor][indices[:4] + indices[5:8]])

2019 ranking Country  ... Foundations of Wellbeing Opportunity
70           10.0   Japan  ...                    91.04       76.26
104           1.0  Norway  ...                    92.32       83.64

[2 rows x 7 columns]

## Maximum

You can get the maximum values of the data frame or a column using the `max` method.  You can also get the row index of the data item that give the maximum value with the `idxmax` method.

By using the `loc` method, you can retrieve multiple entries of the table.  For example, for row numbers $i, j, k$, you can get their row data by `spi2019.loc[[i, j, j]]`.

以下の例が示すように、Social Progress Index について最高値となる国は Norway である。

In [22]:
SPI = 'Social Progress Index'
print(spi2019[SPI].max(), spi2019[SPI].idxmax())
i = spi2019[SPI].idxmax()
spi2019.loc[[i]][indices[:7]]

90.95 104


2019 ranking Country  ... Basic Human Needs Foundations of Wellbeing
104           1.0  Norway  ...             96.89                    92.32

[1 rows x 7 columns]

## Exercise

For each of the following indices, find the member state that gives the maximum and minimum values.

1. Social Progress Index

1. Basic Human Needs

1. Foundations of Wellbeing

1. Opportunity

In [39]:
for i in indices[4:8]:
  h2('Maximum '+i)
  display(spi2019.loc[[spi2019[i].idxmax()]][indices[:4]+[i]])
  h2('Minimum '+i)
  display(spi2019.loc[[spi2019[i].idxmin()]][indices[:4]+[i]])

2019 ranking Country Code  Status  Social Progress Index
104           1.0  Norway  NOR  Ranked                  90.95

2019 ranking      Country Code  Status  Social Progress Index
128         149.0  South Sudan  SSD  Ranked                  24.44

2019 ranking    Country Code  Status  Basic Human Needs
124          27.0  Singapore  SGP  Ranked              98.39

2019 ranking                   Country Code  Status  Basic Human Needs
25         148.0  Central African Republic  CAF  Ranked              23.55

2019 ranking Country Code  Status  Foundations of Wellbeing
104           1.0  Norway  NOR  Ranked                     92.32

2019 ranking      Country Code  Status  Foundations of Wellbeing
128         149.0  South Sudan  SSD  Ranked                     26.28

2019 ranking Country Code  Status  Opportunity
104           1.0  Norway  NOR  Ranked        83.64

2019 ranking      Country Code  Status  Opportunity
128         149.0  South Sudan  SSD  Ranked         14.9

## Exercise

For the four indices in the previous excersize, calculate the mean ([mean](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.mean.html#pandas.DataFrame.mean)) and the standard deviation([std](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.std.html#pandas.DataFrame.std)).

In [51]:
for i in indices[4:8]:
  h2(i)
  print(" Mean", spi2019[i].mean(), "\n Standard Deviation", spi2019[i].std())

 Mean 65.85606666666666 
 Standard Deviation 15.859544747338777


 Mean 74.90684523809517 
 Standard Deviation 18.34200349970229


 Mean 68.32018987341772 
 Standard Deviation 14.571305694591928


 Mean 52.22958083832334 
 Standard Deviation 16.414034156572377


## Exercise

Find the deviation for each country with respect to Social Progress Index.

For the three countries, Norway, Japan, United States, show their Social Progress Index and their deviation.

Ignore members states for which Social Progress Index is given as `NaN`.  Both `mean` and `std` can deal with `NaN` values.  Refer the manual: [mean](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.mean.html#pandas.DataFrame.mean), [std](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.std.html#pandas.DataFrame.std)  (hint: `skipna`).

**Hint**: You may want to use the [merge](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html#pandas.DataFrame.merge) method.

In [68]:
spi_mean = spi2019[SPI].mean()
spi_std = spi2019[SPI].std()

deviation = spi2019[SPI].sub(spi_mean).divide(spi_std)
spi2019["SPI Deviation"] = deviation

mask_JpnNor = (spi2019['Country'] == 'Japan') | (spi2019['Country'] == 'Norway') | (spi2019['Country'] == 'United States')
display(spi2019[mask_JpnNor][indices[:5] + ['SPI Deviation']])

2019 ranking        Country  ... Social Progress Index SPI Deviation
70           10.0          Japan  ...                 88.34      1.417691
104           1.0         Norway  ...                 90.95      1.582261
146          26.0  United States  ...                 83.62      1.120078

[3 rows x 6 columns]